File: 算法/C++/78. Subsets.cpp
class Solution {
public:
    void helper(vector<vector<int>> &ans, vector<int> &nums, vector<int> &subset, int start) {
        ans.push_back(subset);
        for (int i = start; i < nums.size(); i++) {
            subset.push_back(nums[i]);
            helper(ans, nums, subset, i + 1);
            subset.pop_back();
        }
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        //recurrence
        vector<vector<int>> ans;
        vector<int> subset;
        helper(ans, nums, subset, 0);
        return ans;
    }
};

File: 算法/Python/78. Subsets.py
class Solution:
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        # recurrence
        ans = list()
        subset = list()

        def helper(subset, start):
            ans.append(subset[::])
            for i in range(start, len(nums)):
                subset.append(nums[i])
                helper(subset, i+1)
                subset.pop()
        helper(subset, 0)
        return ans
